### PR TITLE
Don't list false positive Go modules

### PR DESCRIPTION
**Pull request by chmeliik**
STONEBLD-1092

Cachi2 used to call `go list -m all` and report all the modules listed
by this command. The command lists lots of false positives - modules
that cachi2 doesn't really download.

Don't use the command anymore, instead get the list of downloaded
modules from `go mod download -json` and add locally replaced modules
from `go list -deps -json all`.

The approach is based on the results of this investigation:
https://github.com/chmeliik/godeps#conclusion

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- [ ] ~Docs updated (if applicable)~
- [ ] ~Docs links in the code are still valid (if docs were updated)~

### CODE REVIEW

**Comment by chmeliik** [comment]
> +18,889 −672

Before you start sharpening your knives, most of that is generated mock data for unit tests :sweat_smile:

File: cachi2/core/package_managers/gomod.py [hunk 1]
```diff
@@ -818,72 +826,83 @@
...
+        )
+        raise UnexpectedFormat(f"vendor/modules.txt: {msg}", solution=solution)
 
-    Note that vendor/modules.txt is fully managed by go. After you call go mod vendor, this file
-    is guaranteed to contain only the content written in it by go.
-    """
-    modules_txt: RootedPath = app_dir.join_within_root("vendor", "modules.txt")
-    module_lines: List[str] = []
-    has_packages = {}
+    def parse_module_line(line: str) -> GoModule:
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py:842]: CodeQL doesn't understand the [NoReturn](https://docs.python.org/3/library/typing.html#typing.NoReturn) type /shakes_fist

File: tests/unit/data/gomod-mocks/generate.sh [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+set -o errexit -o nounset -o pipefail
+
+git clone https://github.com/cachito-testing/gomod-pandemonium
```
Comment (chmeliik) [inline tests/unit/data/gomod-mocks/generate.sh:4]: The unit tests are based around this repo. It exposes the one corner cases of our chosen module identification method - a [locally replaced module](https://github.com/cachito-testing/gomod-pandemonium/blob/a18784b5653c5ca26a2efa7d47143d0274bf5e70/go.mod#L35) which should be listed but isn't (because it's [excluded by build constraints](https://github.com/cachito-testing/gomod-pandemonium/blob/a18784b5653c5ca26a2efa7d47143d0274bf5e70/where-was-i-built/idk.go#L1)).

You can see that when vendoring, this module is reported (as expected), but when not vendoring, it isn't. That's the corner case identified in https://github.com/chmeliik/godeps#conclusion
```diff
❯ diff tests/unit/data/gomod-mocks/resolve_gomod_output.json tests/unit/data/gomod-mocks/resolve_gomod_output_vendored.json  --color
29a30,34
>       "name": "github.com/cachito-testing/gomod-pandemonium/weird",
>       "version": "./weird"
>     },
```
I can also add two integration test cases for this repo (with/without vendoring) if you think it's worth it, or conversely, I can base the unit tests around a less corner-case-y repo :smile:

**Review by chmeliik — COMMENTED** [review]
PTAL - ended up being a rather large refactoring but most of the changes are unit test data

File: tests/unit/data/gomod-mocks/generate.sh [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    echo "generating go_list_deps_threedot.json"
+    go list -deps -json=ImportPath,Module,Standard,Deps ./... > ../go_list_deps_threedot.json
+
+    echo "generating vendor_modules.txt"
+    go mod vendor
+    cp vendor/modules.txt ../vendor_modules.txt
+)
+echo "---------------------------------------------------------------"
+
+echo "please adjust resolve_gomod_output[_vendored].json accordingly"
```
Comment (brunoapimentel) [inline tests/unit/data/gomod-mocks/generate.sh:27]: Took me a minute to figure out what this meant. Different from the integration tests, this generates data that will be used during the test, not the final output that will be used in assertions.

Maybe add a header comment to explain that?

> The generated data will be used as mock output for Go commands, which are called internally by Cachi2. This will may pontentially change the assertion data located in resolve_gomod_output[_vendored].json. The content of these files needs to be manually adjusted.

File: tests/unit/data/gomod-mocks/generate.sh (hunk 2 above)
Comment (chmeliik) [reply tests/unit/data/gomod-mocks/generate.sh:4]: Hmm, that's a good question. Re-generating the data whenever we update Go definitely makes sense, but it's not something we've ever done before (in general, testing different Go versions has never been a thing, most int tests are still on Go 1.14 or something)

I'll add the Makefile command at least, and leave a comment in the "process for updating Go versions" JIRA

File: cachi2/core/package_managers/gomod.py [hunk 4]
```diff
@@ -256,142 +301,111 @@ def _resolve_gomod(
...
 
-        # Top-level packages always use the module version
-        pkg = {"name": pkg_name, "type": "go-package", "version": module_version}
-        packages.append({"pkg": pkg, "pkg_deps": pkg_level_deps})
+        main_pkg = {
+            "type": "go-package",
+            "name": pkg.import_path,
+            "version": main_module_version,
+        }
+        main_packages.append({"pkg": main_pkg, "pkg_deps": pkg_deps})
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py:375]: Cachito the Old will still need the structure, so I figured the changes will be more similar if I keep it here

File: cachi2/core/package_managers/gomod.py [hunk 5]
```diff
@@ -225,29 +268,31 @@ def _resolve_gomod(
 
     run_params = {"env": env, "cwd": app_dir}
 
-    # Collect all the dependency names that are being replaced to later verify if they were
-    # all used
-    replaced_dep_names = set()
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py:230]: Yeah, I filed https://issues.redhat.com/browse/STONEBLD-1132

File: cachi2/core/package_managers/gomod.py [hunk 6]
```diff
@@ -186,16 +189,59 @@ def _find_missing_gomod_files(source_path: RootedPath, subpaths: list[str]) -> l
     return invalid_gomod_files
 
 
-def _resolve_gomod(
-    app_dir: RootedPath, request: Request, tmp_dir: Path, git_dir_path: Optional[Path] = None
-) -> dict[str, Any]:
+class _GolangModel(pydantic.BaseModel):
+    """Attributes automatically get CamelCase aliases to make parsing Golang JSON easier.
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply cachi2/core/package_managers/gomod.py]: TIL, fixed :+1:

File: tests/unit/data/gomod-mocks/generate.sh (hunk 3 above)
Comment (chmeliik) [reply tests/unit/data/gomod-mocks/generate.sh]: Restructured the script a bit, now the message about having to edit the expected data is only printed if there's a git diff in the mocked data

File: tests/unit/data/gomod-mocks/go_list_deps_threedot.json [hunk 7]
```diff
@@ -0,0 +1,6141 @@
...
+	"ImportPath": "github.com/moby/term",
+	"Module": {
+		"Path": "github.com/moby/term",
+		"Version": "v0.0.0-20221205130635-1aeaba878587",
+		"Time": "2022-12-05T13:06:35Z",
+		"Indirect": true,
+		"Dir": "/home/acmiel/go/pkg/mod/github.com/moby/term@v0.0.0-20221205130635-1aeaba878587",
+		"GoMod": "/home/acmiel/go/pkg/mod/cache/download/github.com/moby/term/@v/v0.0.0-20221205130635-1aeaba878587.mod",
+		"GoVersion": "1.18"
+	},
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply tests/unit/data/gomod-mocks/go_list_deps_threedot.json:804]: Ok it does - the main module and locally replaced modules have `Dir` even even when vendoring

[the remark this answers is not in the document]
Comment (brunoapimentel) [reply tests/unit/data/gomod-mocks/go_list_deps_threedot.json:804]: :phew-obama:

**Comment by lkolacek** [comment]
> PTAL - ended up being a rather large refactoring but most of the changes are unit test data

It was still hard to review. :D It looks amazing, thank you Adam! :+1: Great job with "tidying up" gomod code and mocking unit tests.

File: tests/unit/data/gomod-mocks/go_list_deps_threedot.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (lkolacek) [reply tests/unit/data/gomod-mocks/go_list_deps_threedot.json:804]: If I were you, I would not bother that much with it. +1 to use two separate sets of data.

File: tests/integration/test_data/gomod_e2e_test/bom.json [hunk 8]
```diff
@@ -199,16 +199,6 @@
...
-    {
-      "name": "github.com/kr/pty",
-      "type": "library",
-      "version": "v1.1.1"
-    },
-    {
-      "name": "github.com/kr/text",
-      "type": "library",
-      "version": "v0.1.0"
-    },
```
[the remark this answers is not in the document]
Comment (chmeliik) [reply tests/integration/test_data/gomod_e2e_test/bom.json:211]: WDYT, worth the 2 int. test scenarios or not? It's pretty trivial to add them (Long live testdata generation! :tada: :tada: )

File: tests/unit/data/gomod-mocks/non-vendored/go_mod_download.json [hunk 9]
```diff
@@ -0,0 +1,200 @@
+{
+	"Path": "github.com/Microsoft/go-winio",
+	"Version": "v0.6.0",
+	"Info": "/home/acmiel/go/pkg/mod/cache/download/github.com/!microsoft/go-winio/@v/v0.6.0.info",
```
Comment (brunoapimentel) [inline tests/unit/data/gomod-mocks/non-vendored/go_mod_download.json:4]: I know this is only supposed to be mock data for the tests, but I'm worried that everytime we need to re-generate it, the path will change depending on the person/user who did it.

Comment (chmeliik) [reply tests/unit/data/gomod-mocks/non-vendored/go_mod_download.json:4]: Good point, I didn't notice that. Should be easy enough to fix.

File: tests/integration/test_data/gomod_e2e_test/bom.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (brunoapimentel) [reply tests/integration/test_data/gomod_e2e_test/bom.json:211]: What are we expecting to cover with these integration tests exactly? Our local replacement + build constraints corner case?

File: tests/unit/data/gomod-mocks/non-vendored/go_mod_download.json (hunk 9 above)
Comment (chmeliik) [reply tests/unit/data/gomod-mocks/non-vendored/go_mod_download.json]: There was another problem with `Dir` and `GoMod` in `go list -deps -json` output, especially if you want to use `Dir` later

```
"Dir": "/home/acmiel/RedHat/Cachito/cachi2/tests/unit/data/gomod-mocks/gomod-pandemonium",
```

Fixed by setting both GOMODCACHE and the repo path to a tmpdir and replacing the tmpdir with `{tmpdir}` in case someone needs to replace it later

Comment (chmeliik) [reply tests/unit/data/gomod-mocks/non-vendored/go_mod_download.json]: Or maybe better, one `{gomodcache_dir}` placeholder and one `{repo_dir}` placeholder

File: tests/integration/test_data/gomod_e2e_test/bom.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (chmeliik) [reply tests/integration/test_data/gomod_e2e_test/bom.json:211]: All the cases described in https://github.com/cachito-testing/gomod-pandemonium

Currently, we have 2 of them covered:
* locally replaced module in go.mod (https://github.com/cachito-testing/cachito-gomod-local-deps.git)
* test-only dependencies (https://github.com/cachito-testing/retrodep.git)

[the remark this answers is not in the document]
Comment (brunoapimentel) [reply tests/integration/test_data/gomod_e2e_test/bom.json:211]: I'm assuming that using gomod-pandemonium in the unit tests is not enough to have these cases fully tested. In that case, let's add the integration test.

I'm now wondering if the pademonium integration test can replace one of the current integration tests. Maybe `gomod_local_deps`?

[the remark this answers is not in the document]
Comment (chmeliik) [reply tests/integration/test_data/gomod_e2e_test/bom.json:211]: The unit tests mostly checks what our code does with static data, and the integration test would verify if Go really returns that data

But in this case, given how we generate that data, the line is a little blurry. In either case, I'd leave the int. test for a separate PR

**Review by brunoapimentel — APPROVED** [review]

**Review by lkolacek — APPROVED** [review]
